### PR TITLE
feature(api-v2): Generate client test data for ontologies, resources, search, and lists

### LINKED ISSUE #1548: Add v2 clientapi test data for resources, ontologies, and lists
Generate test data listed in https://github.com/dasch-swiss/knora-api-js-lib/blob/54ec4990f56e82e3738aec7b596a5f0dabce4eaa/scripts/v2-test-data-config.json#L2-L49.

Needed by https://github.com/dasch-swiss/knora-api-js-lib/pull/115.

### PR DESCRIPTION
**Pull request by benjamingeer**
This PR incorporates more API v2 routes into the client code generation framework so they can return client test data.

- [x] ontologies route
- [x] resources route
- [x] search route
- [x] lists route

Resolves #1548.

### CODE REVIEW

**Comment by tobiasschweizer** [comment]
In v2 test data for resources, I cannot find JSON data for creation and metadata update ops.

**Comment by tobiasschweizer** [comment]
see https://github.com/dasch-swiss/knora-api-js-lib/blob/9e16288916c2e2a5a66b7d0d9f4448c7dd911438/src/api/v2/resource/resources-endpoint.spec.ts#L96-L648 for the tests I have written so far (test data hard coded)

**Comment by tobiasschweizer** [comment]
I think the the values themselves do not have to be tested thoroughly because this is already done in https://github.com/dasch-swiss/knora-api-js-lib/blob/wip/create-resource/src/api/v2/values/values-endpoint.spec.ts with test data generated by Knora.

**Comment by benjamingeer** [comment]
> In v2 test data for resources, I cannot find JSON data for creation and metadata update ops.

I just included the stuff you asked for 😜:

https://github.com/dasch-swiss/knora-api/pull/1546#issuecomment-562212304

**Comment by benjamingeer** [comment]
Will add test data for the update operations, too.

**Comment by tobiasschweizer** [comment]
> Will add test data for the update operations, too.

Thanks!

**Comment by tobiasschweizer** [comment]
In the meantime I am changing from static to generated v2 test data files. Works already for the ontologies! :-)

My goal is to get rid of static test data at all.

**Comment by tobiasschweizer** [comment]
there is a failing test in  ExceptionHandlerR2RSpec: https://github.com/dasch-swiss/knora-api/pull/1549/checks?check_run_id=341570711#step:6:1310

File: webapi/src/main/scala/org/knora/webapi/routing/v2/ListsRouteV2.scala [hunk 1]
```diff
@@ -19,64 +19,114 @@
...
+class ListsRouteV2(routeData: KnoraRouteData) extends KnoraRoute(routeData) with Authenticator with ClientEndpoint {
+
+    // Definitions for ClientEndpoint
+    override val name: String = "ListsEndpoint"
+    override val directoryName: String = "lists"
+    override val urlPath: String = "lists"
+    override val description: String = "An endpoint for working with Knora lists."
+    override val functions: Seq[ClientFunction] = Seq.empty
+
+    def knoraApiPath: Route = getList ~ getNode
```
Comment (tobiasschweizer) [inline webapi/src/main/scala/org/knora/webapi/routing/v2/ListsRouteV2.scala]: Could you add a comment that this is actually the route definition?

File: webapi/src/main/scala/org/knora/webapi/routing/v2/OntologiesRouteV2.scala [hunk 2]
```diff
@@ -19,85 +19,270 @@
...
+    override val directoryName: String = "ontologies"
+    override val urlPath: String = "ontologies"
+    override val description: String = "An endpoint for working with Knora ontologies."
+    override val functions: Seq[ClientFunction] = Seq.empty
+
     private val ALL_LANGUAGES = "allLanguages"
     private val LAST_MODIFICATION_DATE = "lastModificationDate"
 
-    def knoraApiPath: Route = {
+    def knoraApiPath: Route = dereferenceOntologyIri ~ getOntologyMetadata ~ updateOntologyMetadata ~ getOntologyMetadataForProjects ~
```
Comment (tobiasschweizer) [inline webapi/src/main/scala/org/knora/webapi/routing/v2/OntologiesRouteV2.scala]: Could you add a comment that this is actually the route definition?

File: webapi/src/main/scala/org/knora/webapi/routing/v2/ResourcesRouteV2.scala [hunk 3]
```diff
@@ -127,138 +147,61 @@ class ResourcesRouteV2(routeData: KnoraRouteData) extends KnoraRoute(routeData)
         }
     }
 
+    def knoraApiPath: Route = createResource ~ updateResourceMetadata ~ getResourcesInProject ~
```
Comment (tobiasschweizer) [inline webapi/src/main/scala/org/knora/webapi/routing/v2/ResourcesRouteV2.scala]: Could you add a comment that this is actually the route definition?

File: webapi/src/main/scala/org/knora/webapi/routing/v2/SearchRouteV2.scala [hunk 4]
```diff
@@ -19,32 +19,49 @@
...
+    override val urlPath: String = "search"
+    override val description: String = "An endpoint for searching data in Knora."
+    override val functions: Seq[ClientFunction] = Seq.empty
+
     private val LIMIT_TO_PROJECT = "limitToProject"
     private val LIMIT_TO_RESOURCE_CLASS = "limitToResourceClass"
     private val OFFSET = "offset"
     private val LIMIT_TO_STANDOFF_CLASS = "limitToStandoffClass"
 
+    def knoraApiPath: Route = fullTextSearchCount ~ fullTextSearch ~ gravsearchCountGet ~ gravsearchCountPost ~
```
Comment (tobiasschweizer) [inline webapi/src/main/scala/org/knora/webapi/routing/v2/SearchRouteV2.scala]: Could you add a comment that this is actually the route definition?

**Comment by benjamingeer** [comment]
> there is a failing test in ExceptionHandlerR2RSpec

That's not a failing test, the test is supposed to throw an exception. The GitHub CI build is actually failing here, but I don't know why:

https://github.com/dasch-swiss/knora-api/pull/1549/checks?check_run_id=341570711#step:6:2012

**Comment by benjamingeer** [comment]
> Could you add a comment that this is actually the route definition?

I've added it to all routes.

I added more test requests/responses for the resources route. Do you want update requests for the ontologies route too?

**Comment by tobiasschweizer** [comment]
> I've added it to all routes.

Great, thanks.

> I added more test requests/responses for the resources route. Do you want update requests for the ontologies route too?

Yes, we will need that too. 

Could you also create JSON test data for a resource preview (response to a creation request)?

**Comment by tobiasschweizer** [comment]
The preview JSON data has a higher priority to finish https://github.com/dasch-swiss/knora-api-js-lib/pull/115, so if you need more time for the ontologies and want to do this in a subsequent PR, that's also fine.

**Comment by benjamingeer** [comment]
> if you need more time for the ontologies and want to do this in a subsequent PR

Yes, because I'd like to finish #1509 first.

**Comment by tobiasschweizer** [comment]
When I access http://localhost:3333/clientapi/typescript I now get

{
"error": "akka.http.scaladsl.model.IllegalUriException: Illegal URI reference: Invalid input '/', expected userinfo-char, pct-encoded or '@' (line 1, column 36): http://0.0.0.0:3333resourcespreview/http%3A%2F%2Frdfh.ch%2F0001%2Fa-thing\n                                   ^"
}

**Comment by tobiasschweizer** [comment]
```
"anything:hasDecimal" : {
    "@type" : "knora-api:DecimalValue",
    "knora-api:decimalValueAsDecimal" : {
      "@type" : "x  sd:decimal",
      "@value" : "100000000000000.000000000000001"
    }
  }
```

I think this is too much precision

**Comment by tobiasschweizer** [comment]
Maybe I could look at https://github.com/royNiladri/js-big-decimal or some similar lib.

**Comment by tobiasschweizer** [comment]
if possible, could you also generate data for 

- update a resource's metadata submitting the last modification date
- response to metadata update
- response to resource deletion

**Comment by benjamingeer** [comment]
> "error": "akka.http.scaladsl.model.IllegalUriException

Sorry, stupid mistake.

> update a resource's metadata submitting the last modification date

Yes and it looks like there's a bug, the last modification date should be required.

**Comment by tobiasschweizer** [comment]
When I try to generate client api code I now get:

{"error":"org.knora.webapi.ClientApiGenerationException: Failed to get test data: The requested resource could not be found."}

However, I can get http://localhost:3333/v2/resourcespreview/http%3A%2F%2Frdfh.ch%2F0001%2Fa-thing

File: webapi/src/main/scala/org/knora/webapi/routing/v2/ResourcesRouteV2.scala [hunk 5]
```diff
@@ -376,7 +376,7 @@ class ResourcesRouteV2(routeData: KnoraRouteData) extends KnoraRoute(routeData)
 
     private def getResourcesPreviewTestResponse: Future[SourceCodeFileContent] = {
         for {
-            responseStr <- doTestDataRequest(Get(s"${baseApiUrl}resourcespreview/${SharedTestDataADM.AThing.iriEncoded}"))
+            responseStr <- doTestDataRequest(Get(s"$baseApiUrl/resourcespreview/${SharedTestDataADM.AThing.iriEncoded}"))
```
Comment (tobiasschweizer) [inline webapi/src/main/scala/org/knora/webapi/routing/v2/ResourcesRouteV2.scala]: you are still missing v2: http://0.0.0.0:3333/resourcespreview/http%3A%2F%2Frdfh.ch%2F0001%2Fa-thing

**Comment by benjamingeer** [comment]
> org.knora.webapi.ClientApiGenerationException

Yet another stupid mistake. I think I need to hibernate for the winter.

Can I change the file extensions to `.jsonld` instead of `.json`, or would that be a pain?

File: webapi/src/main/scala/org/knora/webapi/routing/v2/ResourcesRouteV2.scala (hunk 5 above)
Comment (tobiasschweizer) [reply webapi/src/main/scala/org/knora/webapi/routing/v2/ResourcesRouteV2.scala]: Shall I fix it?

Comment (benjamingeer) [reply webapi/src/main/scala/org/knora/webapi/routing/v2/ResourcesRouteV2.scala]: I'm doing it now along with the other stuff you asked for.

**Comment by tobiasschweizer** [comment]
> Can I change the file extensions to .jsonld instead of .json, or would that be a pain?

I experienced some problems earlier with that extension in the past when loading the files in a JS env. But we can give it a try in a separate PR. I am supposed to publish a new version of knora-api-js-lib today for @lrosenth, so I just want to finish resource creation.

File: webapi/src/main/scala/org/knora/webapi/routing/v2/ResourcesRouteV2.scala (hunk 5 above)
Comment (tobiasschweizer) [reply webapi/src/main/scala/org/knora/webapi/routing/v2/ResourcesRouteV2.scala]: Ok, thank you.

**Comment by benjamingeer** [comment]
OK please try it now.

**Comment by tobiasschweizer** [comment]
Ok, that's all I need to finish https://github.com/dasch-swiss/knora-api-js-lib/pull/115, thanks!

I spotted a potential issue for the automatic integration tests in Knora we want to set up (#1554): some of the data is dynamic (such as creation / modification dates) and the tests will fail because the assertions are hard-coded (but they won't fail because of an actual problem we would have to fix).

**Comment by benjamingeer** [comment]
> some of the data is dynamic (such as creation / modification dates)

I think I've fixed this now. Let me know if it's OK to merge.

**Comment by tobiasschweizer** [comment]
> I think I've fixed this now. Let me know if it's OK to merge.

Works :-)

**Comment by benjamingeer** [comment]
> Works :-)

Did you forget to press the "Approve" button, or is there something else I need to do?

**Comment by benjamingeer** [comment]
Thanks for your patience!

**Comment by tobiasschweizer** [comment]
> Thanks for your patience!

No worries!

**Comment by tobiasschweizer** [comment]
Once this is merged, could you update the branches for the modification date and the time value in Knora. I then have to regenerate the client data from these branches to integrate them in the PR is knora-api-js-lib.

**Pull request merged**